src/realm/array.cpp

```diff
@@ -1608,12 +1608,13 @@ MemRef Array::clone(MemRef mem, Allocator& alloc, Allocator& target_alloc)
 
 void Array::do_copy_on_write(size_t minimum_size)
 {
-    // Calculate size in bytes (plus a bit of matchcount room for expansion)
+    // Calculate size in bytes
     size_t array_size = calc_byte_len(m_size, m_width);
-    size_t new_size = std::max(array_size + 64, minimum_size);
-    size_t rest = (~new_size & 0x7) + 1;
-    if (rest < 8)
-        new_size += rest; // 64bit blocks
+    size_t new_size = std::max(array_size, minimum_size);
+    new_size = (new_size + 0x7) & ~size_t(0x7); // 64bit blocks
+    // Plus a bit of matchcount room for expansion
+    if (new_size < max_array_payload - 64)
+        new_size += 64;
```

Comment:
Maybe this was discussed in the CPH office, but since this is a critical issue from so long ago I would request @finnschiermer and/or @rrrlasse to also verify this fix (I can see we had something similar by the `Table_AllocatorCapacityBug` test). In summary, if `copy_on_write()` is called on an array that has a size that is `max_array_payload - 64 < size < max_array_payload` then we will hit an assertion like this `Array::alloc() Assertion failed: value <= max_array_payload [16777272, 16777215]` It can be triggered by users using binary data close to the limit as shown in the added test.